MAVE-NN: learning genotype-phenotype maps from multiplex assays of variant effect
========================================================================
Compatible with Python 3.10-3.12

[![PyPI version](https://badge.fury.io/py/mavenn.svg)](https://badge.fury.io/py/mavenn)
[![Downloads](https://static.pepy.tech/personalized-badge/mavenn?period=total&units=international_system&left_color=black&right_color=blue&left_text=Downloads)](https://pepy.tech/project/mavenn)
[![Documentation Status](https://readthedocs.org/projects/mavenn/badge/?version=latest)](https://mavenn.readthedocs.io/en/latest/?badge=latest)


MAVE-NN enables the rapid quantitative modeling of genotype-phenotype (G-P) maps from the data produced by multiplex assays of variant effect (MAVEs). Such assays include deep mutational scanning (DMS) experiments on proteins, massively parallel reporter assays (MPRAs) on DNA or RNA regulatory sequences, and more. MAVE-NN conceptualizes G-P map inference as a problem in information compression; this problem is then solved by training a neural network using a TensorFlow backend. For installation instructions, tutorials, and documentation, please refer to the MAVE-NN website, https://mavenn.readthedocs.io/. For an extended discussion of this approach and its applications, please refer to our manuscript:

* Tareen, A., Kooshkbaghi, M., Posfai, A. et al. MAVE-NN: learning genotype-phenotype maps from multiplex assays of variant effect. <em>Genome Biol</em> **23**, 98 (2022). https://doi.org/10.1186/s13059-022-02661-7

